### PR TITLE
Answer sheets expanded

### PR DESCRIPTION
**Pull request by DuToit-Hexagon**
Added the register confirmation and declaration to the answer sheets when sent out to teachers or manually downloaded. Also made the email address included in emails dependent on the Django settings

### CODE REVIEW

File: uctMaths/competition/compadmin.py [hunk 1]
```diff
@@ -1261,7 +1263,13 @@ def generate_school_answer_sheets(request, school_list):
     with zipfile.ZipFile(output_stringIO, 'w') as zipf:
         for ischool in school_list:
             output_string=printer_answer_sheet(request, ischool)
-            zipf.writestr(get_answer_sheet_name(ischool), output_string.getvalue())
```
Comment (ianwj5int) [inline uctMaths/competition/compadmin.py:1264]: The PDF is specific to this year's competition. How feasible is it to make this future proof, e.g, a PDF that does not specify the year, or allow a new PDF to be uploaded each year by the competition administrator rather than by a Hexagon developer? The teacher declaration will also fall away if the competition returns to UCT campus in 2023.

File: uctMaths/competition/views.py [hunk 2]
```diff
@@ -596,6 +622,7 @@ def has_results(request, assigned_school = None):
 
 @login_required 
 def answer_sheets(request, assigned_school = None):
+    print(request.user)
```
Comment (ianwj5int) [inline uctMaths/competition/views.py:625]: debug code - remove?

File: uctMaths/competition/compadmin.py (hunk 1 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1264]: The Register Confirmation PDF is in it's own folder in the repo as a pdf, so I think it can be changed by the administrator by just adding a new file. For the declaration, would it work to place the part where it is added to the pdf in a try, finally statement so that it is skipped if the file is deleted, which is presumably what would happen next year?

File: uctMaths/uctMaths/background_tasks.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+from PyPDF2 import PdfFileMerger
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/uctMaths/background_tasks.py:1]: Is this a new library. If so we should add it and all of it's dependencies to req.txt

File: uctMaths/competition/compadmin.py (hunk 1 above)
Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/compadmin.py:1264]: I think we should look into providing a way to upload the Register Confirmation PDF through the admin interface. I don't think we want to change the git repo and redeploy the site every year just to change a PDF.

Alternatively it might be easier to generate the PDF with dates changed? I don't know what you think is easier Du Toit.

File: uctMaths/competition/interface/individual_as_template.html [hunk 4]
```diff
@@ -42,140 +42,140 @@ <h1 style="text-align:center; font-size:18pt; font-family:'Univers',sans-serif;
...
 
-    <table align = "center" width = 85% height = 7%>
-        <tr style="line-height: 125%; text-align:center; font-size:11pt; font-family:'Univers',sans-serif; padding: 3px; vertical-align: middle;">
-          <td width=8% ><b>CODE</b><br>{{code}}</td>
-          <td width=28% ><b>NAME</b><br>{{name}}</td>
-          <td width=28% ><b>SCHOOL</b><br>{{school}}</td>
-          <td width=8% ><b>GRADE</b><br>{{grade}}</td>
-          <td width=28% ><b>VENUE</b><br>{{venue}}</td>
+    <table class = "answersheet" align = "center" width = 85% height = 7%>
+        <tr class = "answersheet" style="line-height: 125%; text-align:center; font-size:11pt; font-family:'Univers',sans-serif; padding: 3px; vertical-align: middle;">
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/interface/individual_as_template.html:46]: I don't think there should be spaces between the `=` here so can you change `class = "answersheet"` to `class="answersheet"`

File: uctMaths/competition/compadmin.py [hunk 5]
```diff
@@ -1288,7 +1296,6 @@ def get_student_answer_sheet(request, student):
         template = get_template('pair_as_template.html')
     else:
         template = get_template('individual_as_template.html')
-    c.update(csrf(request))
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin.py:1291]: Why did you remove this line?

File: uctMaths/competition/compadmin.py [hunk 6]
```diff
@@ -1297,6 +1304,59 @@ def printer_answer_sheet(request, assigned_school=None):
...
+            'invigilator_range':range(10-len(invigilator_list)), 
+            'igrades':range(8,13),
+            'total_num':int(count_pairs*2+count_individuals),
+            'year':year,
+            'invigilators_required':competition_has_invigilator()}
+        #Render the template with the context (from above)
+        template = get_template('printer_entry.html')
+        context = Context(c)
+        html += template.render(context) #Concatenate each rendered template to the html "string"
+
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin.py:1359]: Did you copy paste this whole code block from somewhere? If so can we rather turn this into a function.

File: uctMaths/competition/views.py [hunk 7]
```diff
@@ -75,7 +76,7 @@ def printer_entry_result(request, school_list=None):
         if not (responsible_teacher or alt_responsible_teacher) and not school_list:
             return HttpResponseRedirect('../students/newstudents.html')
         #If the school has an entry
-        elif responsible_teacher:
+        elif responsible_teacher and request:
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/views.py:79]: Why did you split these branches the other branch still has ` c.update(csrf(request))`?

File: uctMaths/competition/compadmin.py (hunk 6 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1359]: I figured as much, but I'm not really comfortable enough with Django to know which folder to place the function in; do you have any ideas?

File: uctMaths/competition/compadmin.py (hunk 5 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1291]: When done from the server side, the request is null, so it breaks completely when csrf is run with None

File: uctMaths/competition/views.py (hunk 7 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/views.py:79]: I honestly can't remember why I did this, I'm testing removing this now

File: uctMaths/competition/compadmin.py (hunk 6 above)
Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/compadmin.py:1359]: looking at they way imports are done. I'd say you should probably refactor this function into `compadmin.py` and then replace this block of code in `views.py` (I think that's where you copied it from...) with a call to the function in `compadmin.py`.

File: uctMaths/competition/compadmin.py (hunk 5 above)
Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/compadmin.py:1291]: I think it would be better to write

``` python
if request is not None:
    c.update(csrf(request))
```

I'm suprised we haven't. We haven't caught this ealier. This would cause the function  `bg_generate_as_grade_distinction` to fail.

File: uctMaths/competition/views.py [hunk 8]
```diff
@@ -45,88 +46,7 @@ def printer_entry_result(request, school_list=None):
...
-
-    result = StringIO.StringIO()
-
-    #Generate the pdf doc
-    pdf = pisa.pisaDocument(StringIO.StringIO(html.encode("UTF-8")), result, encoding='UTF-8')
-    if not pdf.err:
-        return result
-    else:
-        pass #Error handling?
+    return compadmin.generate_answer_sheets(request, school_list)
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/views.py:49]: should `school_list` be `temp_school_list` here?

File: uctMaths/competition/views.py [hunk 9]
```diff
@@ -33,7 +34,7 @@ def printer_entry_result(request, school_list=None):
     #had to easy_install html5lib pisa
     #Create the HttpResponse object with the appropriate PDF headers.
     temp_school_list = []
-    if not school_list: #If not called by the admin
+    if not school_list and not request == None: #If not called by the admin
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/views.py:37]: Who else is calling this function that you need to add `not request == None` here?

File: uctMaths/competition/interface/individual_as_template.html [hunk 10]
```diff
@@ -2,7 +2,7 @@
     <head>
   
         <style>
-        tr {
+        tr.answersheet {
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/interface/individual_as_template.html:5]: BTW, why was this necessary?

Comment (DuToit-Hexagon) [reply uctMaths/competition/interface/individual_as_template.html:5]: The confirmation is also html, so if you combine the forms without setting the styles to be for only the answer sheets the confirmation comes out very strangely

File: uctMaths/competition/views.py (hunk 9 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/views.py:37]: I think I put this in when I was trying to remove the answer sheet generation entirely from compadmin to views, but I tested and it's no longer necessary. I'll remove this now

File: uctMaths/competition/views.py (hunk 8 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/views.py:49]: Yeah, sorry about that

File: uctMaths/competition/compadmin.py (hunk 5 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1291]: I did `if request:`, which also seems to work fine

**Comment by DuToit-Hexagon** [comment]
I added capability for the admin to upload a declaration from the competition screen in the admin gui.

File: uctMaths/competition/interface/pair_as_template.html [hunk 11]
```diff
@@ -42,117 +42,117 @@ <h1 style="text-align:center; font-size:18pt; font-family:'Univers',sans-serif;
...
-          <td width=5% >24</td>
-          <td width=5% >25</td>
-          <td width=5% >26</td>
-          <td width=5% >27</td>
-          <td width=5% >28</td>
-          <td width=5% >29</td>
-          <td width=5% >30</td>
-          <td width=25% >Vraag nommer</td>
+        <tr class="answersheet" style="padding-top: 3px; background-color: lightgray; line-height: 100%;">
+          <td class="answersheet" width=25% >Question number</td>
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/interface/pair_as_template.html:120]: Please, remove this I think it's not needed anymore.

File: uctMaths/uctMaths/background_tasks.py (hunk 3 above)
Comment (yaseen-mowzer-hexagon) [inline uctMaths/uctMaths/background_tasks.py:1]: Can you remove this too please.

File: uctMaths/competition/compadmin.py [hunk 12]
```diff
@@ -1288,25 +1296,121 @@ def get_student_answer_sheet(request, student):
         template = get_template('pair_as_template.html')
     else:
         template = get_template('individual_as_template.html')
-    c.update(csrf(request))
     context = Context(c)
     return template.render(context)
 
+def generate_answer_sheets(request, school_list):
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin.py:1302]: Can you move this back into views.py please. It will make it easier to see what is happening in the diff.

File: uctMaths/competition/reports.py [hunk 13]
```diff
@@ -84,13 +84,14 @@ def send_answer_sheets(school, answer_sheet, cc_admin=False):
...
+    email = EmailMessage(
         "(Do not reply) UCT Mathematics Competition %s Answer Sheets" % (school.name),
         output_string,
-        'UCT Mathematics Competition <%s>'%(settings.DEFAULT_FROM_EMAIL),#from
-        [{"name": "%s" % (compadmin.get_answer_sheet_name(school)), "value": answer_sheet.getvalue(), "type": "application/pdf"}],
+        'UCT Mathematics Competition <%s>'%(settings.DEFAULT_FROM_EMAIL),
         recipient_list
     )
+    email.attach_file(os.path.join(__file__,"..","..","Declaration", str(school) + "_answer_sheets.pdf"))
+    email.send()
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/reports.py:94]: Is this change necessary? Can we do it the old way please?

File: uctMaths/competition/interface/printer_entry.html [hunk 14]
```diff
@@ -207,4 +220,4 @@ <h2>Registered Invigilators</h2>
         </td>
     </tr>
 </table>
-<br><br><br><br>
+<pdf:nextpage />
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/interface/printer_entry.html:223]: What is this for?

File: uctMaths/competition/compadmin_views.py [hunk 15]
```diff
@@ -59,6 +64,47 @@ def upload_results(request):
...
+
+    #Once the user has pressed 'Submit'
+    if request.method == 'POST':
+        #a=request.POST
+        try: #Try receive file from 'Submit' post from user
+            form = UploadDeclarationForm(request.POST, request.FILES)
+            file = request.FILES['upload_file']
+            filepath = os.path.join(__file__,"..","..","Declaration")
+            filename = str(file)
+            if not filename == 'Declaration.pdf':
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin_views.py:81]: why don't we just set `filename` to `Declaration.pdf` and ignore what the user called the file.

File: uctMaths/competition/compadmin.py [hunk 16]
```diff
@@ -1288,25 +1296,121 @@ def get_student_answer_sheet(request, student):
...
     else:
-        pass #Error handling?
+        register_result = generate_answer_sheets(request, [assigned_school])
+    answer_sheets_pdf = pisa.pisaDocument(StringIO.StringIO(answer_sheets_html.encode("UTF-8")), answer_sheets_result, encoding='UTF-8')
+    merger = PdfFileMerger()
+    merger.append(register_result)
+    merger.append(answer_sheets_result)
+    merger.append(os.path.join(__file__,"..","..","Declaration","Declaration.pdf"))
+    merger.write(os.path.join(__file__,"..","..","Declaration","temp_answer_sheets.pdf"))
+    pdf = open(os.path.join(__file__,"..","..","Declaration","temp_answer_sheets.pdf"))
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin.py:1407]: You need to close this and close PdfFileMerger

**Review by yaseen-mowzer-hexagon — CHANGES_REQUESTED** [review]

File: uctMaths/competition/reports.py (hunk 13 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/reports.py:94]: This doesn't work, background_tasks has request as None, which I think isn't allowed in functions in views.py as the traceback ends on the function call into views.py

Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/reports.py:94]: I don't really understand, there is no `request` anywhere in this function?

File: uctMaths/competition/compadmin.py [hunk 17]
```diff
@@ -1288,25 +1296,121 @@ def get_student_answer_sheet(request, student):
...
+        register_result = generate_answer_sheets(request, assigned_school)
     else:
-        pass #Error handling?
+        register_result = generate_answer_sheets(request, [assigned_school])
+    answer_sheets_pdf = pisa.pisaDocument(StringIO.StringIO(answer_sheets_html.encode("UTF-8")), answer_sheets_result, encoding='UTF-8')
+    merger = PdfFileMerger()
+    merger.append(register_result)
+    merger.append(answer_sheets_result)
+    merger.append(os.path.join(__file__,"..","..","Declaration","Declaration.pdf"))
+    merger.write(os.path.join(__file__,"..","..","Declaration","temp_answer_sheets.pdf"))
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/competition/compadmin.py:1406]: Oh yeah, I forgot to mention. Can you create a temporary file/dir instead of hardcoding the path. Hard coding the path might cause issues if two requests happen at the same time.

File: uctMaths/competition/reports.py (hunk 13 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/reports.py:94]: Whoops, wrong message

Comment (DuToit-Hexagon) [reply uctMaths/competition/reports.py:94]: I mean wrong conversation

File: uctMaths/competition/compadmin.py (hunk 12 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1302]: This doesn't work, background_tasks has request as None, which I think isn't allowed in functions in views.py as the traceback ends on the function call into views.py
(This should be here)

File: uctMaths/competition/compadmin.py (hunk 17 above)
Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/compadmin.py:1406]: Actually, even better, I think just using a `StringIO` here will work.

Comment (DuToit-Hexagon) [reply uctMaths/competition/compadmin.py:1406]: Sure, no worries

File: uctMaths/competition/interface/printer_entry.html (hunk 14 above)
Comment (DuToit-Hexagon) [reply uctMaths/competition/interface/printer_entry.html:223]: Is it necessary to add the <br>s back in? They cause an extra blank page to appear when a school has a lot of entries

File: uctMaths/uctMaths/background_tasks.py [hunk 18]
```diff
@@ -55,8 +55,8 @@ def bg_generate_school_answer_sheets(school_id):
     if not pdf:
         print("%s: Unable to get answer sheets!" % (current_time()))
         return
-    
     send_answer_sheets(school, pdf, True)
+    pdf.close()
```
Comment (yaseen-mowzer-hexagon) [inline uctMaths/uctMaths/background_tasks.py:59]: You don't need to close pdf, since it's a `StringIO` and `StringIO` doesn't need to be closed.

File: uctMaths/competition/compadmin.py (hunk 5 above)
Comment (yaseen-mowzer-hexagon) [reply uctMaths/competition/compadmin.py:1291]: Actually this is fine, we don't need CSRF.

**Review by yaseen-mowzer-hexagon — APPROVED** [review]
